src/state/slices/stakingDataSlice/selectors.ts

```diff
@@ -169,14 +218,152 @@ export const selectRewardsByAccountSpecifier = createDeepEqualOutputSelector(
   }
 )
 
+export const selectAllRewardsByDenom = createDeepEqualOutputSelector(
+  selectStakingDataByAccountSpecifier,
+  selectDenom,
+  (stakingData, denom): Record<string, chainAdapters.cosmos.Reward[]> => {
+    if (!stakingData || !stakingData.rewards) return {}
+
+    const rewards = stakingData.rewards.reduce(
+      (acc: Record<string, chainAdapters.cosmos.Reward[]>, current: ValidatorReward) => {
+        if (!acc[current.validator.address]) {
+          acc[current.validator.address] = []
+        }
+
+        acc[current.validator.address].push(
+          ...current.rewards.filter(x => ASSET_ID_TO_DENOM[x.assetId] === denom)
+        )
+
+        return acc
+      },
+      {}
+    )
+
+    return rewards
+  }
+)
+
 export const selectRewardsAmountByDenom = createSelector(
   selectRewardsByAccountSpecifier,
+  selectValidatorAddress,
   selectDenom,
-  (rewardsByAccountSpecifier, denom): string => {
+  (rewardsByAccountSpecifier, validatorAddress, denom): string => {
     if (!rewardsByAccountSpecifier.length) return ''
 
     const rewards = rewardsByAccountSpecifier.find(rewards => rewards.denom === denom)
 
     return rewards?.amount || ''
   }
 )
+
+export const selectAllValidators = createDeepEqualOutputSelector(
+  selectStakingData,
+  stakingData => stakingData.byValidator
+)
+
+export const selectSingleValidator = createSelector(
+  selectStakingData,
+  selectValidatorAddress,
+  (stakingData, validatorAddress) => {
+    return stakingData.byValidator[validatorAddress] || null
+  }
+)
+
+export const selectNonloadedValidators = createSelector(
+  selectStakingDataByAccountSpecifier,
+  selectAllValidators,
+  (stakingData, allValidators): string[] => {
+    if (!stakingData) return []
+    const initialValidatorsAddresses = stakingData.delegations
+      ? stakingData.delegations.map(x => x.validator.address)
+      : []
+    if (stakingData.undelegations) {
+      initialValidatorsAddresses.push(...stakingData.undelegations.map(x => x.validator.address))
+    }
+
+    if (stakingData.rewards) {
+      initialValidatorsAddresses.push(...stakingData.rewards.map(x => x.validator.address))
+    }
+
+    const uniqueValidatorAddresses = [...new Set(initialValidatorsAddresses)]
+    return uniqueValidatorAddresses.filter(x => allValidators[x] === undefined)
+  }
+)
+
+export const selectUndelegationsAmountByValidatorAddress = memoize(
+  (
+    allUndelegationsEntries: Record<string, chainAdapters.cosmos.UndelegationEntry[]>,
+    validatorAddress: string
+  ) => {
+    return get(
+      allUndelegationsEntries,
+      validatorAddress,
+      [] as chainAdapters.cosmos.UndelegationEntry[]
+    )
+      .reduce((acc: BigNumber, undelegationEntry: chainAdapters.cosmos.UndelegationEntry) => {
+        acc = acc.plus(bnOrZero(undelegationEntry.amount))
+        return acc
+      }, bnOrZero(0))
+      .toString()
+  },
+  (
+    allUndelegationsEntries: Record<string, chainAdapters.cosmos.UndelegationEntry[]>,
```

Comment:
@NeOMakinG Hey there, there was a bug in rewards and undelegationsAmount calculation. The probleme was with memoize that takes first argument as key to store in cache (https://lodash.com/docs/4.17.15#memoize). Basically since the first array (allRewards for example) is the same for every call, it would store the first value computed (here 0 most of the time). I fixed this by adding a resolver as a param of the memoize function.

Reply:
Ok, looks fine to me! nice catch